code/go/0chain.net/sharder/blockstore/cache.go

```diff
@@ -0,0 +1,296 @@
+package blockstore
+
+// Cache is a simple implementation using simple lru for storing block's minimal data.
+// Sharder should provide config for cache which includes basically path, and its size.
+// Each read cache will be stored uncompressed into the cache path.
+// For cache is full, and size of latest read block is n then blocks are removed from blocks such
+// that total removed size is >= n
+
+import (
+	"container/list"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/0chain/common/core/logging"
+
+	"0chain.net/core/viper"
+)
+
+const (
+	DefaultCacheBufferSize = 100
+	KB                     = 1024
+	MB                     = 1024 * KB
+	GB                     = 1024 * MB
+)
+
+type cacher interface {
+	Write(hash string, data []byte) error
+	Read(hash string) ([]byte, error)
+}
+
+// cache manages blocks cache
+type cache struct {
+	// path to store uncompressed blocks
+	path string
+	// sizeLimit limit of total blocks size in the path
+	sizeLimit int64
+
+	// size current total blocks size in the path
+	size int64
+
+	lru      *lru
```

Comment:
use cache.LRU as main implementation

Reply:
I checked the cache.LRU if you mean from `cache` package in `core` directory.
It is limiting for the required implementation and cannot be modified without affecting other packages that requires it.

Reply:
Cache is SSD disk. To manage blocks in a directory cache.LRU is not enough.
Which is why local implementation is used.

Reply:
your LRu implementation is in memory LRU cache , am I missing something?

Reply:
https://github.com/0chain/0chain/blob/2260920e4ee7e530258dc50c8681c1062d4aa198/code/go/0chain.net/sharder/blockstore/cache.go#L182:L211

Reply:
The main cache is SSD disk which is managed by LRU cache in Ram memory.
Managing two caches requires low level access to the lru and cache.LRU cannot provide that facility.

Check [this use case](https://github.com/0chain/0chain/blob/2260920e4ee7e530258dc50c8681c1062d4aa198/code/go/0chain.net/sharder/blockstore/cache.go#L147).

Reply:
I can't get you, we already have LRU in the project please use it, what is wrong with it?

Reply:
@dabasov Don't understand what you are not getting.
First thing is that sharder can specify any amount of SSD cache size(better maybe to have it between 500MB-2GB). Minimum I've set is 500MB. Note that this is SSD disk space where the blocks are stored uncompressed.

The size of block is not fixed. Some blocks can have maximum number of transactions and some can have minimum number thus making block size different from each other.

The LRU cache in memory depends on the size of block as well.
So if SSD cache size is 100 MB and block size is for example constant to 1MB then LRU cache can have maximum of 100 key-value data.
Same SSD cache size with block size of 500 KB, LRU cache  can have maximum of 200 key-value data.

The LRU that you are suggesting to use has `size` parameter to initialize cache but size cannot be determined beforehand.

The other thing is; When sharder tries to push a block to SSD cache and finds that SSD cache is full then it will have to remove last used blocks. So we need to have access to doubly-linked-list to get block-id(hash) so as to delete last used block. For example, if sharder is trying to push block of size 1MB and last block is 500KB size and second last block is 600 KB then it will remove two blocks that were last used. This feature is not possible in the LRU that you mentioned.

cache.LRU is independent cache for itself.
The LRU cache that I wrote is co-dependent on SSD cache size and block size.

FYI Before writing own LRU, I've thoroughly checked if I could use existing LRU logic but it did not fit the requirement.

Reply:
calculate size and initialize cache with it, is it a problem? if you need different one, reinitialize it. please, do not repeat already implemented code, we try to use common code as much as possible

```
The LRU that you are suggesting to use has size parameter to initialize cache but size cannot be determined beforehand.
```
please reference where it is dependent? I'll think how to rewrite it

Reply:
If common LRU cache does preoccupy memory then it is fine. We can calculate cache size considering minimum size of block.
The other important requirements are:

1. To be able get last element in the cache so as to delete it in LRU cache and SSD cache as well.
2. When new key-value is inserted into LRU cache it should return if [insertion is new](https://github.com/0chain/0chain/blob/enhancement%2Fblocks-management/code/go/0chain.net/sharder/blockstore/cache.go#L208-L218)(a boolean value).